include/libcyphal/presentation/subscriber_impl.hpp

```diff
@@ -0,0 +1,289 @@
+                if (!tmp_buffer)
+                {
+                    return false;
+                }
+                const auto data_size = context.buffer.copy(0, tmp_buffer.get(), context.buffer.size());
+
+                const auto* const data_raw = static_cast<const void*>(tmp_buffer.get());
+                const auto* const data_u8s = static_cast<const std::uint8_t*>(data_raw);  // NOSONAR cpp:S5356 cpp:S5357
+                const nunavut::support::const_bitspan bitspan{data_u8s, data_size};
+
+                return deserialize(message, bitspan);
+            }
+
+            template <typename Message, typename Subscriber>
+            static void deserializeMsgOnceForManySubs(Context& context)
+            {
+                CETL_DEBUG_ASSERT(context.next_node != nullptr, "");
+                constexpr auto this_deserializer_fn = &deserializeMsgOnceForManySubs<Message, Subscriber>;
+                CETL_DEBUG_ASSERT(context.next_node->deserializer_.function == this_deserializer_fn, "");
+
+                // Deserialize the message from the buffer - only once!
+                //
+                Message    message{};
+                const bool got_message = tryDeserializeMessage(message, context);
+
+                // Enumerate all nodes with the same deserializer, and deliver the message to them.
+                // It's important to do it even in case of deserialization failure -
+                // to "consume" nodes and make sure that enumeration is progressing.
+                //
+                while (auto* const curr_node = context.next_node)
+                {
+                    if (curr_node->deserializer_.function != this_deserializer_fn)
+                    {
+                        // We've reached the end of the list of nodes with the same deserializer.
+                        // A different deserializer will take care of the next node.
+                        break;
+                    }
+
+                    // "Consume" current node by preparing the context for the next iteration.
+                    // It's important to do it before the callback b/c its activity may modify what is next.
+                    context.next_node = curr_node->getNextInOrderNode();
+
+                    // Notify only those subscribers that were created before the message was received.
+                    // This is to avoid nondeterministic delivery of messages to subscribers that were created
+                    // after the message was sent to one of callbacks.
+                    //
+                    if (got_message && (context.approx_now > curr_node->creation_time_))
+                    {
+                        // This is safe downcast b/c we know that the `curr_node` is of type `Subscriber`.
+                        // Otherwise, the `deserializer_` would be different from `this_deserializer`.
+                        auto* const subscriber = static_cast<Subscriber*>(curr_node);
```

Comment:
A very interesting design but it will break one day when we added the queue depth parameterization to the Subscriber (and perhaps some other compile-time policies). Generally, depending on a concrete type of the user-facing type is not a good idea, I think.

Reply:
>will break one day when we added the queue depth parameterization
so, let's don't add it ;) - see my above answer about storage adapters